### PR TITLE
initial module checkin

### PR DESCRIPTION
**Pull request by Yongqu**
This is initial runtime support for module ( parser change at 
https://github.com/Microsoft/ChakraCore/pull/293).
    This PR contain the basic support for host integration, and it can act as basic bootstrap to
enable edge support later. 
    Current es6 module spec does not clarify the dependency module loading sequence, and module 
loader spec is not finalized yet. At this time, we will allow browser host to start from scripttype='module' 
tag, and allow host to asynchronously fetch child module sources. This is in effect adding a small state 
machine between ParseModule and ModuleDeclarationInstantiation. Once all modules are parsed and 
initialized, the script engine notifies the host, and ModuleEvaluation of top level module will start 
asynchronously, initiated from the host. 
    There are still a lot of unimplemented features, including experimental JSRT APIs that match 
execution model, namespace object and ResolveExports, error handling etc. Also this change is 
developed in parallel to the parser change it will be integrated with the parser part later.

### CODE REVIEW

**Comment by bmeck** [comment]
Please note that the WHATWG Loader and the Node loader have intrinsic differences and the Loader will likely require host environment setup outside of the VM. These problems are explained in https://github.com/nodejs/node-eps/pull/3 which holds the Node a companion proposal to the WHATWG proposal about how Node needs to instrument things.

**Comment by bmeck** [comment]
reference to another potential change in the spec (at JS level, not Loader) : https://github.com/tc39/ecma262/issues/368#issuecomment-180611264

**Comment by Yongqu** [comment]
@bmeck Thanks for the info and I'll take a look. We do want to support both es6 and node interaction. @bterlson

File: lib/Runtime/Base/ScriptContext.h [hunk 1]
```diff
@@ -68,6 +69,19 @@ enum ExternalJitData
     ExternalJitData_CustomExternalObjectOperations
 };
 
+enum LoadScriptFlag
+{
+    LoadScriptFlag_None         = 0x0,
+    LoadScriptFlag_Expression   = 0x1,                  // the script returns a value.
+    LoadScriptFlag_disableDeferredParse = 0x2,          // do not defer-parse the code.
```
Comment (boingoing) [inline lib/Runtime/Base/ScriptContext.h]: Want to make casing consistent between these flags?

File: lib/Runtime/Language/SourceTextModuleRecord.cpp [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    TempArenaAllocatorObject* SourceTextModuleRecord::EnsureTempAllocator()
+    {
+        if (tempAllocatorObject == nullptr)
+        {
+            tempAllocatorObject = scriptContext->GetThreadContext()->GetTemporaryAllocator(L"Module");
+        }
+        return tempAllocatorObject;
+    }
+
+
```
Comment (boingoing) [inline lib/Runtime/Language/SourceTextModuleRecord.cpp]: Extra whitespace

File: lib/Runtime/Language/SourceTextModuleRecord.cpp [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        Assert(!wasEvaluated);
+        parseTree = nullptr;
+        requestModuleList = nullptr;
+        importRecordList = nullptr;
+        localExportRecordList = nullptr;
+        indirectExportRecordList = nullptr;
+        starExportRecordList = nullptr;
+        childrenModuleSet = nullptr;
+        parentModuleList = nullptr;
+        AllocatorDelete(ArenaAllocator, tempAllocatorObject->GetAllocator(), parser);
```
Comment (boingoing) [inline lib/Runtime/Language/SourceTextModuleRecord.cpp]: Does AllocatorDelete set parser to nullptr? Should we do that here?

File: lib/Runtime/Base/ScriptContext.cpp [hunk 4]
```diff
@@ -1589,37 +1589,44 @@ namespace Js
...
-    JavascriptFunction* ScriptContext::LoadScript(const wchar_t* script, SRCINFO const * pSrcInfo, CompileScriptException * pse, bool isExpression, bool disableDeferredParse, bool isByteCodeBufferForLibrary, Utf8SourceInfo** ppSourceInfo, const wchar_t *rootDisplayName, bool isLibraryCode, bool disableAsmJs, bool isSourceModule)
+    ParseNode* ScriptContext::ParseScript(Parser* parser,
+        const byte* script,
+        size_t cb,
+        SRCINFO const * pSrcInfo,
+        CompileScriptException * pse,
+        Utf8SourceInfo** ppSourceInfo,
+        const wchar_t *rootDisplayName,
+        LoadScriptFlag loadScriptFlag,
+        uint* sourceIndex)
```
Comment (boingoing) [inline lib/Runtime/Base/ScriptContext.cpp]: Cool, this script flag is being tested by code in module-syntax.js, right? Assume there is no way to have unit tests for the module linking code yet.

**Comment by boingoing** [comment]
Big change looks good to me!

**Comment by pleath** [comment]
Does this change contain the compile-time API that takes an import symbol and returns the module index and symbol index?

**Comment by Yongqu** [comment]
@pleath no that's part of ModuleDeclarationInitialization and it is not implemented yet.

**Comment by pleath** [comment]
Okay. Thanks. What's here LGTM.

**Pull request merged**